src/validation.cpp

```diff
@@ -4432,62 +4432,19 @@ bool ChainstateManager::LoadBlockIndex()
         std::sort(vSortedByHeight.begin(), vSortedByHeight.end(),
                   CBlockIndexHeightOnlyComparator());
 
-        // Find start of assumed-valid region.
-        int first_assumed_valid_height = std::numeric_limits<int>::max();
-
-        for (const CBlockIndex* block : vSortedByHeight) {
-            if (block->IsAssumedValid()) {
-                auto chainstates = GetAll();
-
-                // If we encounter an assumed-valid block index entry, ensure that we have
-                // one chainstate that tolerates assumed-valid entries and another that does
-                // not (i.e. the background validation chainstate), since assumed-valid
-                // entries should always be pending validation by a fully-validated chainstate.
-                auto any_chain = [&](auto fnc) { return std::any_of(chainstates.cbegin(), chainstates.cend(), fnc); };
-                assert(any_chain([](auto chainstate) { return chainstate->reliesOnAssumedValid(); }));
-                assert(any_chain([](auto chainstate) { return !chainstate->reliesOnAssumedValid(); }));
-
-                first_assumed_valid_height = block->nHeight;
-                LogPrintf("Saw first assumedvalid block at height %d (%s)\n",
-                        first_assumed_valid_height, block->ToString());
-                break;
-            }
-        }
-
         for (CBlockIndex* pindex : vSortedByHeight) {
             if (m_interrupt) return false;
-            if (pindex->IsAssumedValid() ||
+            // If we have an assumeutxo-based chainstate, then the snapshot
+            // block will be a candidate for the tip, but it may not be
+            // VALID_TRANSACTIONS (eg if we haven't yet downloaded the block),
+            // so we special-case the snapshot block as a potential candidate
+            // here.
+            if (pindex == GetSnapshotBaseBlock() ||
```

Comment:
In commit "Fix initialization of setBlockIndexCandidates when working with multiple chainstates" (768690b7ce551cd403f8e2a099372915f6022ad4)

It looks like there may be a  bug in this code because this `pindex == GetSnapshotBaseBlock()` special case will add the snapshot block to `setBlockIndexCandidates` whether or not it has data, but there is also an assert in `FindMostWorkChain` which asserts that blocks in `setBlockIndexCandidates` do have data:

https://github.com/bitcoin/bitcoin/blob/c5a63ea56f8347139bd84e1669b378ecfb234c3c/src/validation.cpp#L2914 

I think it's not possible to hit this assert for the snapshot chainstate, but it may be possible to hit this assert for the background chainstate.

For the snapshot chainstate, I think the assert won't be reached because after the snapshot is loaded the `Chainstate::LoadChainTip` function should immediately set the chain tip to the snapshot block index:

https://github.com/bitcoin/bitcoin/blob/c5a63ea56f8347139bd84e1669b378ecfb234c3c/src/validation.cpp#L4153

so the `m_chain.Contains(pindexTest)` check before the assert will be false and the assert will not run:

https://github.com/bitcoin/bitcoin/blob/c5a63ea56f8347139bd84e1669b378ecfb234c3c/src/validation.cpp#L2913-L2914

But for the background chainstate, it seems possible assert could be hit if `FindMostWorkChain` is called before the snapshot block is downloaded.

In any case, this special `pindex == GetSnapshotBaseBlock()` case seems overbroad. It seems like it should only apply to the snapshot chainstate, not the background chainstate, and I'm not actually sure why it is necessary to have at all if `Chainstate::LoadChainTip` is already going to set the snapshot block as the snapshot chainstate tip.

I think the interaction between these checks and the checks inside the `TryAddBlockIndexCandidate` are also unnecessarily confusing, and it would be nice whether there is a bug or not to move all the move all the checks inside `TryAddBlockIndexCandidate` and call it unconditionally. This should make it easeir to avoid adding the snapshot block the background chainstate candidate set before it is downloaded. It would also be great to drop the snapshot block special case entirely if it is not needed.

Reply:
> It looks like there may be a bug in this code because this pindex == GetSnapshotBaseBlock() special case will add the snapshot block to setBlockIndexCandidates whether or not it has data, but there is also an assert in FindMostWorkChain which asserts that blocks in setBlockIndexCandidates do have data:
> 
> [bitcoin/src/validation.cpp](https://github.com/bitcoin/bitcoin/blob/c5a63ea56f8347139bd84e1669b378ecfb234c3c/src/validation.cpp#L2914)
> Line 2914 in [c5a63ea](https://github.com/bitcoin/bitcoin/commit/c5a63ea56f8347139bd84e1669b378ecfb234c3c)
>  ```cpp
>  assert(pindexTest->HaveTxsDownloaded() || pindexTest->nHeight == 0); 
>  ```
> I think it's not possible to hit this assert for the snapshot chainstate, but it may be possible to hit this assert for the background chainstate.

If I'm reading `FindMostWorkChain()` correctly, this assert _won't_ be hit for a background chainstate, because the second clause of this condition 
```cpp
while (pindexTest && !m_chain.Contains(pindexTest)) {
```
will exclude the snapshot base block from consideration without data (since it is not in the background chainstate's `m_chain` until it is actually downloaded, and so has data). So as far as I can tell, there's no bug here - but I'm glad you're double checking this; the `setBlockIndexCandidates` code is very confusing indeed, and I'm amenable to any way we can make it more easily understandable.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/27746#discussion_r1321815541

The way I'm reading e code:

https://github.com/bitcoin/bitcoin/blob/c5a63ea56f8347139bd84e1669b378ecfb234c3c/src/validation.cpp#L2913-L2914

If a snapshot is loaded and `pindexTest` points to the snapshot block and the snapshot block is not yet downloaded, and `m_chain` is the background chain, then `m_chain.Contains(pindexTest)` will be false, so the while loop condition will be true, and the assert could be be hit.

You seem to be saying that `m_chain.Contains(pindexTest)` will be false, so the while condition _won't_ be hit?

Reply:
I'm sorry, my previous comment was more or less incorrect. Another reason why the assert you mention doesn't pose a problem is because, confusingly, `HaveTxsDownloaded()` is just an alias for `nChainTx != 0`. Since we manually set nChainTx for the snapshot base block (https://github.com/bitcoin/bitcoin/pull/27596/commits/315586f25b9ec0ed0c807d8c4455e6a73a3c4990) this check (again, very confusingly) passes. At some point long ago I made a note somewhere about fixing that method's misleading name...

So the long story short is that `FindMostWorkChain()` eventually works for the background chainstate because it detects, correctly, that there are blocks with missing data between the IBD tip and the snapshot base, and correctly rewinds `m_chain` back to the actual IBD tip. I think when @sdaftuar first proposed this change I was confused about why the `LoadBlockIndex()` simplification could actually work, and this is the reason.

Reply:
Thanks, I missed that HaveTxsDownloaded was checking `nChainTx` not `nTx`.

I guess the main thing I'm confused about now is why this `pindex == GetSnapshotBaseBlock() ||` special case is needed at all? If `LoadChainTip()` already sets the snapshot block to be the snapshot tip, why does the snapshot block need to be added to the candidate set?

Reply:
> I guess the main thing I'm confused about now is why this pindex == GetSnapshotBaseBlock() || special case is needed at all? If LoadChainTip() already sets the snapshot block to be the snapshot tip, why does the snapshot block need to be added to the candidate set?

I'm not sure if there's a more fundamental reason, but various assertions fail if this case is stripped out of LoadBlockIndex(); e.g. the `!setBlockIndexCandidates.empty()` assertion fails during `PruneBlockIndexCandidates()` if this special case is not included.

The true purpose of setBlockIndexCandidates has always been somewhat mysterious to me, but the most essential use seems to be in FindMostWorkChain(), to supply candidates for activation in `ActivateBestChain()`.

Reply:
I've added some documentation for `HaveTxsDownloaded()` (https://github.com/bitcoin/bitcoin/pull/27596/commits/9f9ffb655e4236f8d567b961102f53c5933e8aa8), and I'm happy to add more for the confusing process discussed above.